sql/moz-fx-data-shared-prod/udf_js/bootstrap_percentile_ci/udf.sql

```diff
@@ -0,0 +1,212 @@
+/*
+Calculates a confidence interval using an efficient bootstrap sampling technique for a
+given percentile of a histogram.
+
+See:
+- https://mozilla-hub.atlassian.net/browse/DS-2128
+- https://engineering.atspotify.com/2022/03/comparing-quantiles-at-scale-in-online-a-b-testing/
+
+Users must specify an array of percentiles of interest as the first parameter and a
+histogram struct as the second.
+*/
+
+CREATE OR REPLACE FUNCTION udf_js.bootstrap_percentile_ci(
+  percentiles ARRAY<INT64>,
+  histogram STRUCT<values ARRAY<STRUCT<key FLOAT64, value FLOAT64>>>,
+  metric STRING
+)
+RETURNS ARRAY<STRUCT<
+    metric STRING,
+    statistic STRING,
+    point FLOAT64,
+    lower FLOAT64,
+    upper FLOAT64,
+    parameter STRING
+>> DETERMINISTIC
+LANGUAGE js OPTIONS ( library=['gs://moz-data-science-bochocki-bucket/bq_udf_library/qbinom.js'])
+AS
+"""
+  function numericSort(arr) {
+    arr.sort(function(a, b) {return a - b;});
+  }
+
+  function validatePercentile(percentile) {
+    if (percentile < 0 || percentile > 100) {
+      throw "percentile must be a value between 0 and 100";
+    }
+  }
+  
+  function getQuantileSamples(percentile, histogram, n_samples) {
+    validatePercentile(percentile);
+
+    /* expand the histogram */
+    var full_histogram = histogram.map(bin => Array(bin.value).fill(parseFloat(bin.key))).reduce((prev, curr) => prev.concat(curr));
```

Comment:
Just noticing that currently we compute the normalized sum of the histograms and pass them into the forkknife percentile UDF. This wouldn't work here (since `fill` on floats will fail). I am not 100% sure why we compute the normalized sums in the first place, but it seems like with this we can just pass the histograms into the UDFs and not worry about it? Or will this have some implications?

Reply:
Good catch! The previous version normalized the histogram twice I think. Once in the UDF [here](https://github.com/mozilla/bigquery-etl/blob/4ed4bd07be3281c3571720c19172c4acee117489/sql/moz-fx-data-shared-prod/udf_js/jackknife_percentile_ci/udf.sql#L25), and once (I _think_) for the UDF input, [based on the templating here](https://github.com/mozilla/opmon/blob/3291261a49d01efff00a55aedf1e0abb996a1fb7/opmon/templates/statistics_query.sql#L3).

This proposed approach skips the normalization step by simply using the location of values in a sorted array (e.g. `1` is the median value in the array `[0, 1, 1, 2]`). The result is the same, but the proposed approach has better time complexity (it's just a lookup instead of performing the normalization and looping through the histogram).

The simplicity comes with the memory cost of having to reshape the histogram into an array, but the proposed approach requires the data to be in an array form anyway, so we're really just taking advantage of the shape of the data.

One positive side-effect of this is that we could modify this UDF (or create a slightly different one) to accept a scalar array as input. In that case, we wouldn't have to convert scalar columns to histograms before computing confidence intervals, we could just use the columns as-is.

Reply:
To answer your question more directly, we'll need to make sure that we don't normalize the histograms that get passed in.